app/Services/FileImportService.php

```diff
@@ -0,0 +1,275 @@
+                }
+                $csvData[] = $data;
+            }
+
+            if ($fileImport->num_rows == -1) {
+                $fileImport->num_rows = count($csvData);
+                $fileImport->save();
+            }
+
+            for ($i = $fileImport->progress; $i < count($csvData); $i++) {
+                $data = $csvData[$i];
+                if (! $this->isValidDate($data[7])) {
+                    $this->saveFailedImport($fileImport, $data, $user, 'Date must be in the format YYYY-MM-DD');
+
+                    $fileImport->progress = $i + 1;
+                    $fileImport->save();
+
+                    continue;
+                }
+
+                $transactionType = strtolower(trim($data[2]));
+                if (in_array($transactionType, [TransactionType::EXPENSE->value, TransactionType::INCOME->value])) {
+                    try {
+                        $this->importTransaction($data, $transactionType, $user);
+                    } catch (FileImportException $e) {
+                        $this->saveFailedImport($fileImport, $data, $user, $e->getMessage());
+                        Log::error($e);
+                    } catch (Exception $e) {
+                        $this->saveFailedImport($fileImport, $data, $user, 'An error occurred while importing this transaction');
+                        Log::error($e);
+                    }
+                } elseif ($transactionType == '+Transfer') {
+                    if (isset($csvData[$i + 1]) && $csvData[$i + 1][2] == '-Transfer') {
+                        try {
+                            $this->importTransfer($data, $csvData[$i + 1], $transactionType, $user);
+                        } catch (FileImportException $e) {
+                            $this->saveFailedImport($fileImport, $data, $user, $e->getMessage());
+                            Log::error($e);
+                        } catch (Exception $e) {
+                            $this->saveFailedImport($fileImport, $data, $user, 'An error occurred while importing this transaction');
+                            $this->saveFailedImport($fileImport, $csvData[$i + 1], $user, 'An error occurred while importing this transaction');
+                            Log::error($e);
+                        } finally {
+                            $i += 1;
+                        }
+                    } else {
+                        $this->saveFailedImport($fileImport, $data, $user, 'Corresponding -Transfer transaction not found');
+                    }
+                } else {
+                    $this->saveFailedImport($fileImport, $data, $user, 'Invalid transaction type');
+                }
+
+                $fileImport->progress = $i + 1;
+                $fileImport->save();
+            }
+
+            $fileImport->refresh();
+
+            // delete the file after reading
+            unlink($path);
+            fclose($handle);
+            ImportComplete::dispatch($fileImport);
```

Comment:
Ummm :  https://csv.thephpleague.com/

Looks complex to me and we probably should not be doing that, e nothing this here so you create another issue to migrate this implementation

Reply:
https://github.com/trakli/webservice/issues/30